test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -63,64 +126,95 @@ class ApiUsers::PermissionsControllerTest < ActionController::TestCase
+    should "raise an exception if the application is retired" do
       sign_in create(:superadmin_user)
 
-      application = create(:application)
-      api_user = create(:api_user)
-      create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
-      create(:access_token, resource_owner_id: api_user.id, application:)
-
-      get :edit, params: { api_user_id: api_user, application_id: application }
+      application = create(:application, retired: true)
+      permission = create(:supported_permission, application:)
+      api_user = create(:api_user,
+                        with_signin_permissions_for: [application],
+                        with_permissions: { application => [permission.name] })
+      create(:access_token, application:, resource_owner_id: api_user.id)
 
-      assert_response :success
+      assert_raises(ActiveRecord::RecordNotFound) do
+        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [application.id] } }
+      end
     end
 
-    should "raise an exception if the user does not have an access token for the application" do
+    should "raise an exception if the user cannot be found" do
       application = create(:application)
-      api_user = create(:api_user)
 
       current_user = create(:superadmin_user)
       sign_in current_user
 
       assert_raises(ActiveRecord::RecordNotFound) do
-        get :edit, params: { api_user_id: api_user, application_id: application }
+        patch :update, params: { api_user_id: "unknown-id", application_id: application, application: { supported_permission_ids: %w[id] } }
       end
     end
-  end
 
-  context "#update" do
-    should "prevent unauthenticated users" do
+    should "raise an exception when attempting to update with a revoked access token" do
+      sign_in create(:superadmin_user)
+
       application = create(:application)
       api_user = create(:api_user)
+      create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
 
-      patch :update, params: { api_user_id: api_user, application_id: application }
+      assert_raises(ActiveRecord::RecordNotFound) do
+        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
+      end
+    end
 
-      assert_not_authenticated
+    should "include applications with revoked access tokens when there is at least one non-revoked access token" do
+      sign_in create(:superadmin_user)
+
+      application = create(:application, with_non_delegatable_supported_permissions: %w[permission])
+      api_user = create(:api_user)
+      create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
+      create(:access_token, resource_owner_id: api_user.id, application:)
+
+      permission = application.supported_permissions.find_by(name: "permission")
```

Comment:
What do you think about moving this (and other tests) to the other style where we assign the permission to a variable on creation, rather than using the transient then finding?

Reply:
I.e. this change: https://github.com/alphagov/signon/pull/3079/commits/21e7a0ee76dbeec4f16a32f31c9c6d073e4a91ec